pandas/core/frame.py

```diff
@@ -8559,6 +8580,54 @@ def blk_func(values):
                 out[:] = coerce_to_dtypes(out.values, df.dtypes)
             return out
 
+        def array_func(values):
+            if isinstance(values, ExtensionArray):
+                return values._reduce(name, skipna=skipna, **kwds)
+            else:
+                return op(values, skipna=skipna, **kwds)
+
+        # all other options with axis=0 are done column-array-wise
+        if axis == 0:
+
+            def _constructor(df, result, index=None):
+                index = index if index is not None else df.columns
+                if len(result):
+                    return df._constructor_sliced(result, index=index)
+                else:
+                    # set correct dtype for empty result
+                    dtype = "bool" if filter_type == "bool" else "float64"
+                    return df._constructor_sliced(result, index=index, dtype=dtype)
+
+            def _reduce_columns(df, op):
+                result = [op(arr) for arr in df._iter_column_arrays()]
+                return _constructor(df, result)
+
+            df = self
+            if numeric_only is True:
+                df = _get_data(axis_matters=True)
+
+            if numeric_only is not None:
+                return _reduce_columns(df, array_func)
+            else:
+                # need to catch and ignore exceptions when numeric_only=None
+                try:
+                    return _reduce_columns(df, array_func)
+                except TypeError:
+                    # if column-wise fails and numeric_only was None, we try
+                    # again but removing those columns for which it fails
+                    result = []
+                    indices = []
+                    for i, arr in enumerate(df._iter_column_arrays()):
+                        try:
+                            res = array_func(arr)
+                        except Exception:
+                            pass
+                        else:
+                            result.append(res)
+                            indices.append(i)
+
+                    return _constructor(df, result, index=df.columns[indices])
+
         if not self._is_homogeneous_type:
             # try to avoid self.values call
 
```

Comment:
This block (using `frame_apply`) could be removed now, since this is only called when `axis=0`, which is now already handled above.

We *might* want to limit the new code above to `if axis == 0 and self._mgr.any_extension_types:`, and then keep the block here for backwards compatibility (or eg only remove when doing a major version bump pandas 2.0).